core/crypto/src/key_file.rs

```diff
@@ -22,10 +43,259 @@ impl KeyFile {
+        Ok(PasswordInput(buf_str))
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct SodiumoxideSecretBox(secretbox::Nonce);
+
+//https://docs.rs/sodiumoxide/0.2.0/src/sodiumoxide/newtype_macros.rs.html#257-285
+impl Default for SodiumoxideSecretBox {
+    fn default() -> Self {
+        SodiumoxideSecretBox::new()
+    }
+}
+
+impl SodiumoxideSecretBox {
+    pub fn new() -> Self {
+        let nonce = secretbox::gen_nonce();
+        SodiumoxideSecretBox(nonce)
+    }
+
+    //need to be tested
+    pub fn gen_key(bytes: &[u8]) -> Result<Key, Error> {
+        Key::from_slice(bytes).map(Ok).unwrap_or_else(|| {
+            Err(Error::new(
+                ErrorKind::InvalidInput,
+                "Failed to create key from slice".to_string(),
+            ))
+        })
+    }
+}
+
+trait Operation {
+    fn encrypt(&self, m: &[u8], k: &Key) -> Vec<u8>;
+    fn decrypt(&self, c: &[u8], k: &Key) -> Vec<u8>;
+}
+
+impl Operation for SodiumoxideSecretBox {
+    fn encrypt(&self, m: &[u8], k: &Key) -> Vec<u8> {
+        secretbox::seal(m, &self.0, &k)
+    }
+
+    fn decrypt(&self, c: &[u8], k: &Key) -> Vec<u8>{
+        secretbox::open(c, &self.0, &k).unwrap()
+    }
+}
+
+// pub fn perform_encryption(key_file : &mut KeyFile,
+//     key_store_path: &Path, home_dir: &Path,
+//     ) -> Result<(), Error> {
+//         key_file.write_to_file(&key_store_path);
+    
+//     if home_dir.join("config.json").exists() {
+//         //add password corrrectness validation to from_encrypted_file_with_password
+//         let input = PasswordInput::enter_from_console();
+//         KeyFile::from_encrypted_file_with_password(&key_store_path, input)?;
+//     } else {
+//         let input = PasswordInput::enter_from_console();//ask for input of password 
+//         key_file = key_file.write_to_encrypted_file(&key_store_path, input)?;
+//     }
+
+//     Ok(())
+// }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test] 
+    fn test_encrypt_decrypt() {
+        //encryption
+        for key_type in vec![KeyType::ED25519, KeyType::SECP256K1] {
+            let file_name = match key_type {
+                KeyType::ED25519 => "ED25519.txt".to_string(),
+                KeyType::SECP256K1 => "SECP256K1.txt".to_string(),
+            };
+            let secret_key = SecretKey::from_seed(key_type, "test");
+            let public_key  = secret_key.public_key();
+            let account_id = String::from("test");
+            let mut key_file = KeyFile {
+                account_id: account_id,
+                public_key: public_key,
+                secret_key: secret_key.clone(),
+                key_hash: KeyHash::default(),
+                ssb: SodiumoxideSecretBox::default(),
+            };
+            let key_store_path = Path::new(&file_name);
+            let input = "qwerty".to_string();
+            //encryption
+            assert!(key_file.write_to_encrypted_file(&key_store_path, &input));
```

Comment:
`write_to_encrypted_file` re-generates nonce that was created with `SodiumoxideSecretBox::default()`.